### PR TITLE
Add Tempo integration

### PR DESCRIPTION
**Pull request by domiin50**
FineTracker app uses vlsso for login and Jira operations. Currently, it
is unable to use it for submitting worklogs to Tempo or another Tempo
operations.

Add whole Tempo integration for proxying Tempo requests. Integration
functions and flow are inspirated by Jira integration.

Add columns into DB for Tempo Credentials.

To vlsso authorization flow add Tempo authorization by OAuth 2.0.
Handle Tempo server side bugs with query paramters and special
characters within authorization.

Open /api/tempo/* route for listening to all Tempo requests.

As a result, after authorization, vlsso can be used as proxy also for
Tempo requests.

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -17,5 +17,10 @@ JIRA_CONSUMER_PRIVATE_KEY="[base64-encoded-private-key]"
 JIRA_DOMAIN=https://vacuum.atlassian.net
 
 # EMS
-EMS_API_KEY=
-EMS_DOMAIN=
+EMS_API_KEY=ApiKey
```
Comment (hleumas) [inline .env.template]: Let's start with basic stuff. This commit has no reason to change anything related to EMS, so don't do that.

File: src/authorize.js [hunk 2]
```diff
@@ -14,7 +15,7 @@ export function login(req, res) {
   }
   const redirect_uri = url.req2uri(req, {
     pathname: r.oauth,
-    query: {url: req.query.url},
+    query: {url: decodeURIComponent(req.query.url)},
```
Comment (hleumas) [inline src/authorize.js]: Again, how is this related to Tempo?

File: src/authorize.js [hunk 3]
```diff
@@ -44,11 +45,11 @@ export async function oauth(req, res) {
   const authSession = await getAuthSession(sso_access_token)
   const accountId = await jira.requestAccountId({
     token: authSession.jira_access_token,
-    secret: authSession.jira_access_secret
+    secret: authSession.jira_access_secret,
```
Comment (hleumas) [inline src/authorize.js]: How is this change related to Tempo?

File: src/tempo.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  const {code: authorizationToken, error} = req.query
+  if (error) return null
+  return await tokenRequest('authorization_code', {code: authorizationToken, redirect_uri: redirectUri})
+}
+
+export async function requestNewAccessToken(refreshToken) {
+  return await tokenRequest('refresh_token', {refresh_token: refreshToken})
+}
+
+export async function request(
```
Comment (hleumas) [inline src/tempo.js]: Is there a reason why this isn't implemented the same way the github proxying is?

File: src/authorize.js (hunk 2 above)
Comment (domiin50) [reply src/authorize.js]: It is. Tempo has a problem with handling special characters.
It does need to get redirect URL decoded in the query params.

**Comment by domiin50** [comment]
Resolves issue #2.

File: src/authorize.js (hunk 3 above)
Comment (domiin50) [reply src/authorize.js]: I just tried to fix eslint warnings that came in my way while implementing Tempo.
I thought they are just minors that will not bother. In the next commit, I will change it back.

Comment (hleumas) [reply src/authorize.js]: Fixing eslint is fine, but do that in separate commit.

File: src/authorize.js (hunk 2 above)
Comment (hleumas) [reply src/authorize.js]: I don't understand.

File: src/authorize.js [hunk 5]
```diff
@@ -65,10 +66,57 @@ export async function oauthJira(req, res) {
...
   )
+  checkTempo(req, res)
+}
+
+function getTempoCallback(req, ssoAccessToken, redirectUrl) {
+  /*
+  BUGS IN TEMPO API:
+  Tickets on Tempo Support have been already opened.
+
+  Unable to pass multiple query parameters in callback auth url into Tempo API.
```
Comment (hleumas) [inline src/authorize.js]: Why don't we use JSON.stringify then?

Comment (domiin50) [reply src/authorize.js]: Problem with handling one of the brackets and `"`. I had really a problem to find a suitable separator that would be handled properly to send all the needed parameters as one in the query. And also that could not be used in any of the sent parameters.

Comment (hleumas) [reply src/authorize.js]: This is not a correct solution, `redirectUrl` after decoding might contain characters that cause problems for you, you simply can't be sure of that.

On the other hand, you can always use `JSON.stringify` and further encode it using BASE64, which should be pretty safe (I guess).

File: src/authorize.js [hunk 6]
```diff
@@ -65,10 +66,58 @@ export async function oauthJira(req, res) {
     jiraCredentials.jiraId,
     jiraCredentials.accessCredentials,
   )
+  checkTempo(req, res)
+}
+
+function getTempoCallback(req, ssoAccessToken, redirectUrl) {
```
Comment (hleumas) [inline src/authorize.js]: Let's rename this function to `tempoCallback`.

I don't understand why it accepts `req` as argument. I would expect it taking only `ssoAccessToken` and `url`, same as jira callback and github callback do.

File: src/authorize.js [hunk 7]
```diff
@@ -48,7 +49,7 @@ export async function oauth(req, res) {
   })
 
   if (accountId == null) await jira.authorize(res, callback)
-  else res.redirect(url.set(req.query.url, {query: {code: sso_access_token}}))
+  else checkTempo(req, res, sso_access_token)
```
Comment (hleumas) [inline src/authorize.js]: This function does too much. I'd expect explicit `tempo.authorize(res, tempoCallback())` call here.

File: src/authorize.js [hunk 8]
```diff
@@ -65,10 +66,58 @@ export async function oauthJira(req, res) {
...
+  */
+  const query = {ssoAccessToken, url: decodeURIComponent(redirectUrl)}
+  const encodedQuery = Buffer.from(JSON.stringify(query)).toString('base64')
+  return url.req2uri(req, {
+    pathname: r.oauthTempo,
+    query: {params: encodedQuery},
+  })
+}
+
+async function checkTempo(req, res, ssoAccessToken = req.query.sso_access_token) {
```
Comment (hleumas) [inline src/authorize.js]: This function does too much. Actually, I am not even sure we need it at all. Do we?

Comment (hleumas) [reply src/authorize.js]: `authSession` is already retrieved in `oauth` call, `authorize` call should be done explicitly in the `oauth` function, `requestNewAccessToken` can be moved there too.

Why do we call `storeTempoCredentials`?

Comment (domiin50) [reply src/authorize.js]: The flow is the same as in Jira auth.
Make a request to check actual access credentials. The Tempo has no endpoint for something like account info with access credentials usage, so there is used request for refreshing the access credentials. If the previous credentials were valid, store the new ones and continue to the redirect. If they were not, it is needed to authorize.
Again, this function has 2 use cases. It could be called after oAuthJira or if there is no need to authorize Jira, the workflow continues and this function is called immediately.

The whole workflow is the same, just after Jira auth processing, there is no redirect back to the App, but it is added a new block for Tempo auth. After Tempo there is this redirect, that was after Jira. It was also on the two places in the code because it could have two scenarios in processing every auth: It is needed to be authorized or not.

Comment (domiin50) [reply src/authorize.js]: This function has the same workflow as checking Jira in `oauth` method:
```
const callback = url.req2uri(req, {
  pathname: r.oauthJira,
  query: {sso_access_token, url: req.query.url},
})

const authSession = await getAuthSession(sso_access_token)
const accountId = await jira.requestAccountId({
  token: authSession.jira_access_token,
  secret: authSession.jira_access_secret,
})

if (accountId == null) await jira.authorize(res, callback)
else checkTempo(req, res, sso_access_token)
```
Here it is extracted to a function because it could be called from two diff places.
Make sense?

Comment (domiin50) [reply src/authorize.js]: What I see a possible modification is to change the checking function to not retrieve new access credentials and store them to the DB every time but to make a request to some another random endpoint and authorize just when the credentials expired or are somehow changed. With this actual `requestNewAccessToken` function it is a much smaller chance for credentials expiration since they are every time renewed without authorization.

File: src/tempo.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
+import config from './config'
+import fetch from 'node-fetch'
+import {URLSearchParams} from 'url'
+import {map} from 'lodash'
+
+export function authorize(res, callback) {
+  const authorizationUrl = `${config.jira.domain}/plugins/servlet/ac/io.tempo.jira/oauth-authorize/?access_type=tenant_user`
+  const url = `${authorizationUrl}&client_id=${config.tempo.clientId}&redirect_uri=${callback}`
+  res.redirect(url)
+}
```
Comment (domiin50) [inline src/tempo.js]: The following change requested by @hleumas during the meeting:
Proper encoding of URL and its query params when making requests during Tempo auth.

File: src/tempo.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+async function tokenRequest(grantType, authorization) {
+  const {domain, clientId, clientSecret} = config.tempo
+  const searchParams = new URLSearchParams()
+  const params = {
+    grant_type: grantType,
+    client_id: clientId,
+    client_secret: clientSecret,
+    ...authorization,
+  }
+  map(params, (value, key) => searchParams.append(key, value))
```
Comment (domiin50) [inline src/tempo.js]: The following change requested by @hleumas during the meeting:
Use just one consistent way in the project to encode the URL query params.

File: src/authorize.js [hunk 11]
```diff
@@ -43,23 +44,29 @@ export async function oauth(req, res) {
     sso_access_token: ssoAccessToken,
     ...githubCredentials,
   })
-
   await ensureAllAuthorizations(req, res, ssoAccessToken)
 }
 
-async function ensureAllAuthorizations(req, res, token) {
-  const session = await getAuthSession(token)
+async function ensureAllAuthorizations(req, res, ssoAccessToken) {
```
Comment (hleumas) [inline src/authorize.js]: I do get that you were fixing bugs after refactor, but that doesn't explain why you change lot of other stuff such as this function.

File: src/authorize.js [hunk 12]
```diff
@@ -43,23 +44,29 @@ export async function oauth(req, res) {
...
 
-  await ensureJiraAuthorization(session, req, res, token, url)
-  && await ensureTempoAuthorization(session, req, res, token, url)
-  && res.redirect(u.set(req.query.url, {query: {code: token}}))
+  ;(await ensureJiraAuthorization(session, req, res, ssoAccessToken, url)) &&
+    (await ensureTempoAuthorization(session, req, res, ssoAccessToken, url)) &&
+    res.redirect(u.set(url, {query: {code: ssoAccessToken}}))
 }
 
+function getCallback(req, path, ssoAccessToken, url) {
```
Comment (hleumas) [inline src/authorize.js]: Or this function.

File: src/authorize.js [hunk 13]
```diff
@@ -43,23 +44,29 @@ export async function oauth(req, res) {
...
+  return u.req2uri(req, {
+    pathname: path,
+    query: {sso_access_token: ssoAccessToken, url},
+  })
+}
 
-async function ensureJiraAuthorization(session, req, res, token, url) {
+//JIRA Authorization
+async function ensureJiraAuthorization(session, req, res, ssoAccessToken, url) {
```
Comment (hleumas) [inline src/authorize.js]: Or this.

File: src/authorize.js [hunk 14]
```diff
@@ -95,44 +97,33 @@ export async function oauthJira(req, res) {
   await ensureAllAuthorizations(req, res, req.query.sso_access_token)
 }
 
-function tempoCallback(req, sso_access_token, url) {
```
Comment (hleumas) [inline src/authorize.js]: I am confused you deleted this. Wasn't this kind of the point? That tempo really needed a special treatment?

Comment (domiin50) [reply src/authorize.js]: They have already fixed this. It is not needed anymore.

**Comment by hleumas** [comment]
Whatever, fix conflicts and merge.
